From 839215d35fdf6f33f87fdf30ebf960aa2414111e Mon Sep 17 00:00:00 2001
From: Christophe Fergeau <cfergeau@redhat.com>
Date: Fri, 15 Jan 2016 16:35:52 +0100
Subject: [PATCH] ovirt: Use sso-token when set in .vv file

Starting with oVirt 4.0, this replaces the jsessionid field for
automatic authentication with oVirt instances for REST communication.

(cherry picked from commit 4f3733294028d9b1f7c7135eb299436509845bb4)
---
 man/remote-viewer.pod    | 11 +++++++++--
 src/ovirt-foreign-menu.c | 25 ++++++++++++++++++++++---
 2 files changed, 31 insertions(+), 5 deletions(-)

diff --git a/man/remote-viewer.pod b/man/remote-viewer.pod
index 1358f1b..6b69aa3 100644
--- a/man/remote-viewer.pod
+++ b/man/remote-viewer.pod
@@ -309,8 +309,15 @@ GUID of the oVirt virtual machine to connect to.
 
 =item C<jsessionid> (string)
 
-Value to set the 'jsessionid' cookie to. Setting this authentication cookie to a valid value
-will allow to interact with the oVirt REST API without being asked for credentials.
+Value to set the 'jsessionid' cookie to. With oVirt 3.6, setting this
+authentication cookie to a valid value will allow to interact with the oVirt
+REST API without being asked for credentials.
+
+=item C<sso-token> (string)
+
+Value to set the 'Authorization' header to. With oVirt 4.0 or newer, setting
+this authentication header to a valid value will allow to interact with the
+oVirt REST API without being asked for credentials.
 
 =item C<ca> (string)
 
diff --git a/src/ovirt-foreign-menu.c b/src/ovirt-foreign-menu.c
index ca0ac7f..cd5fdc6 100644
--- a/src/ovirt-foreign-menu.c
+++ b/src/ovirt-foreign-menu.c
@@ -835,6 +835,7 @@ OvirtForeignMenu *ovirt_foreign_menu_new_from_file(VirtViewerFile *file)
     gboolean admin;
     char *ca_str = NULL;
     char *jsessionid = NULL;
+    char *sso_token = NULL;
     char *url = NULL;
     char *vm_guid = NULL;
     GByteArray *ca = NULL;
@@ -842,15 +843,22 @@ OvirtForeignMenu *ovirt_foreign_menu_new_from_file(VirtViewerFile *file)
     url = virt_viewer_file_get_ovirt_host(file);
     vm_guid = virt_viewer_file_get_ovirt_vm_guid(file);
     jsessionid = virt_viewer_file_get_ovirt_jsessionid(file);
+    sso_token = virt_viewer_file_get_ovirt_sso_token(file);
     ca_str = virt_viewer_file_get_ovirt_ca(file);
     admin = virt_viewer_file_get_ovirt_admin(file);
 
-    if ((url == NULL) || (vm_guid == NULL) || (jsessionid == NULL)) {
-        g_debug("ignoring [ovirt] section content as URL, VM GUID or jsessionid"
+    if ((url == NULL) || (vm_guid == NULL)) {
+        g_debug("ignoring [ovirt] section content as URL, VM GUID"
                 " are missing from the .vv file");
         goto end;
     }
 
+    if ((jsessionid == NULL) && (sso_token == NULL)) {
+        g_debug("ignoring [ovirt] section content as jsessionid and sso-token"
+                " are both missing from the .vv file");
+        goto end;
+    }
+
     proxy = ovirt_proxy_new(url);
     if (proxy == NULL)
         goto end;
@@ -862,9 +870,19 @@ OvirtForeignMenu *ovirt_foreign_menu_new_from_file(VirtViewerFile *file)
 
     g_object_set(G_OBJECT(proxy),
                  "admin", admin,
-                 "session-id", jsessionid,
                  "ca-cert", ca,
                  NULL);
+    if (jsessionid != NULL) {
+        g_object_set(G_OBJECT(proxy),
+                     "session-id", jsessionid,
+                     NULL);
+    }
+    if (sso_token != NULL) {
+        g_object_set(G_OBJECT(proxy),
+                     "sso-token", sso_token,
+                     NULL);
+    }
+
     menu = g_object_new(OVIRT_TYPE_FOREIGN_MENU,
                         "proxy", proxy,
                         "vm-guid", vm_guid,
@@ -874,6 +892,7 @@ end:
     g_free(url);
     g_free(vm_guid);
     g_free(jsessionid);
+    g_free(sso_token);
     g_free(ca_str);
     if (ca != NULL) {
         g_byte_array_unref(ca);
